readers/multishard.cc

```diff
@@ -810,6 +811,13 @@ future<> shard_reader_v2::do_fill_buffer() {
                             mutation_reader::forwarding fwd_mr) {
                     return lifecycle_policy->create_reader(std::move(s), std::move(permit), pr, ps, std::move(ts), fwd_mr);
                 });
+                if (this_shard_id() == coordinator_shard) {
+                    // Having two permits on the same shard in the same reader
+                    // tree, is a recipe for deadlocks. Release the base resources
+                    // of the multishard reader's permit, so they can be
+                    // transferred to the shard reader below.
+                    _permit.release_base_resources();
+                }
```

Comment:
Shouldn't `release_base_resources()` be called in `multishard_combining_reader_v2` (in the constructor?) instead of `shard_reader_v2::do_fill_buffer()`?

If we only call it in `fill_buffer()`, isn't it still possible to have the deadlock, just a more involved variant?

For example, imagine there are 10 multishard reads (started simultaneously) coordinated by shard 0, and 10 coordinated by by shard 1. But the multishard readers of shard 0 start by reading some token range belonging to shard 1, while the multishard readers on shard 1 start by reading some token range belonging to shard 0.

Then, won't it be the case that the system will deadlock, because all 10/10 resources on both shards are occupied by multishard readers, and their sub-readers keep waiting for admission on the other shard forever? And `release_base_resources()` is never called, since — as done by this commit — the call only happens when a multishard reader starts reading its own shard?

Reply:
More generally speaking, it doesn't really matter whether the two permits are on the same shard or not. Once you have a dependency cycle between semaphores (where "A depends on B" means that the program can request a permit from semaphore B while holding a permit from A), you have a can of worms. This commit only prevents cycles of length 1, which doesn't seem like enough.

Reply:
I wanted `release_base_resources()` to be as close to `obtain_reader_permit()` as possible, so there is a reasonable chance that the resources are transferred. But you are right that this creates a window for deadlocks, so I guess I'll have to move it to the constructor.